### PR TITLE
Make icons part of the theme

### LINKED ISSUE #501: [Themes] Make icons part of the theme
We should use the system icon scheme to better integrate with the look and feel.

### PR DESCRIPTION
**Pull request by plata**
fixes #501 

I don't quite understand when the JavaFX objects (e.g. ViewLibrary) are created. Currently it seems to happen before the ThemeManager is initialized. Therefore, always the default icons are used. Could use some help here.

### CODE REVIEW

**Comment by qparis** [comment]
ThemeManager should not be a singleton. The singleton pattern should be prohibited

**Comment by plata** [comment]
Yes. I just did it as a quick solution until we have the bean. It's not even new in this PR.

**Comment by qparis** [comment]
Can you fix it? It will be a good start to solve your problem

**Comment by plata** [comment]
I would like to. But I'm not sure if I can (I don't really know spring, beans etc.). Maybe I'll give it a try.

I was thinking about something else: Wouldn't it be better to specify the icon path for every icon in the CSS? Even if I get the ThemeManager bean working, the icon would never be reloaded. So if the theme is changed, the icons would still stay the same. If the icons are specified in the CSS, this should work automatically.

**Comment by qparis** [comment]
Any screenshot?

Anyway, I insist on the fact that the theme manager should be a bean and support multiple files. We cannot workaround this problem by trying to have a single file

**Comment by plata** [comment]
themed icons

I agree with the bean. However, I would prefer not do do it with this PR. It will be big enough already (I must change some other stuff before this can be merged.

**Comment by qparis** [comment]
What other stuff? This should really be a priority

**Comment by plata** [comment]
E.g. fix the category icons (they are missing currently because they are not yet part of the theme).

**Comment by qparis** [comment]
They should not be directly handled in CSS because we must be able to provide a default icon in the repository side. 

You really need to make ThemeManager a bean. Therefore, you'll rely on a strong architecture and be able to do much more stuff than you can do now

**Comment by plata** [comment]
If the categories are can not have a theme, it will look like crap because two different icon styles are mixed up.

**Comment by qparis** [comment]
Yep but on the other hand, the categories cannot be fixed

**Comment by plata** [comment]
I don't see a problem. It's not like they change all the time.

**Comment by plata** [comment]
themed app categories

**Comment by qparis** [comment]
An non official repository should be able to add a category, this is very important

**Comment by plata** [comment]
Ok. We could think about a fallback for situations like this.

**Comment by qparis** [comment]
This is a question of consistency. The concept of category is on an ApplicationSource. Therefore, this component should be able to provide all the necessary data

**Comment by plata** [comment]
We cannot provide the theme in the ApplicationSource. That doesn't make any sense.

**Comment by qparis** [comment]
In this case we should go even further and make application miniatures part of the theme ;-)

The icon should be in the AppSource and optionally overwritten by the theme IMHO (exactly as it is done for gnome / kde categories)

**Comment by plata** [comment]
The miniatures are not really icons but pictures so that's ok.

I tried to figure out in the code whether or not the icon is available in the theme. But that doesn't seem to work somehow.

**Comment by plata** [comment]
Not exactly what I want but it has the same effect...

In this picture I removed the category "Custom" from the list of themed categories:
fallback

**Comment by qparis** [comment]
Are you sure that the category icons are well aligned?

**Comment by qparis** [comment]
I really think that we need to handle this differently :-/ I find the current approach not really clean

**Comment by plata** [comment]
Well, if you have a better idea, fell free to propose.

**Comment by qparis** [comment]
I have identified two responsibilities 

First, the theme manager should be able to provide any resource on demand.
Second, it should be able to notify some components when some resource are changed.

So we need 
1- a fetchResource(String name) method
2- a refresh method
2- a set of event handlers called when method is called

This component can be used basically anywhere, and should be initialized before the views. It needs to be a Spring bean so that it is instanciated in the right time

What do you think?

**Comment by plata** [comment]
I'm not sure about the second responsibility: This would require that either the ThemeManager knows everything which could be affected by a theme change or components have to subscribe somehow.

Also, I couldn't find anything similar on the web. So I'm not sure if that's the way JavaFX is supposed to be used (I found quite some examples defining icons in the CSS).

**Comment by plata** [comment]
Maybe the easiest thing would be: Store the icon path (not a `byte[]`) in the CategoryDTO. Then we can just set `-fx-background-image` to that path and we're done.

**Comment by qparis** [comment]
don't use the path, it will no longer work for webservice or class path implementation. However, you can try to use base64 in the CSS file

**Comment by plata** [comment]
How do I use base64?

**Comment by qparis** [comment]
In HTML, it looks like this:
http://stackoverflow.com/questions/17090571/is-there-a-way-to-set-background-image-as-a-base64-encoded-image-in-javascript

**Comment by plata** [comment]
It says: `WARNING: unknown protocol: data`

**Comment by plata** [comment]
Can't the ApplicationSource just give a proper URL?

**Comment by qparis** [comment]
Well I'm not 100% that javafx css supports base64 background images.
If it is not the case, we should not use CSS at it is a solution that does not fit our need.
Better to redesign a solution than twisting something that does not work for us

**Comment by qparis** [comment]
The problem of handling the image in another URL is that we should also handle all the asynchronous stuff required to access it

**Comment by plata** [comment]
Shouldn't the default theme handling do this already? I mean after all the images you specify in the CSS must be loaded as well somehow.

**Comment by qparis** [comment]
In the CSS yep, but javafx is doing it for you. If you go from a byte[] to a URL, you need to implement it

**Comment by qparis** [comment]
Oh sorry my bad!

**Comment by qparis** [comment]
Can you try with a HTTP URL?

**Comment by plata** [comment]
I tried. Doesn't seem to be any problem at all.
The also do it here: <https://stackoverflow.com/questions/10518458/javafx-create-custom-button-with-image>

**Comment by qparis** [comment]
Ok. We just need to ensure that we can do it with classpath

**Comment by plata** [comment]
Doesn't it use classpath by default?

**Comment by qparis** [comment]
It uses git (local repository) by default

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/apps/ViewApps.java [hunk 1]
```diff
@@ -78,10 +84,28 @@ public void showAvailableApps() {
      */
     public void populate(List<CategoryDTO> categories) {
         Platform.runLater(() -> {
+            Set<String> themedCategories = Sets.newHashSet(
+                    "custom",
```
Comment (qparis) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/apps/ViewApps.java]: Why not generalizing it?

**Comment by plata** [comment]
I was talking about the URL resolution. Or does it look in the jar only?

**Comment by qparis** [comment]
I don't understand what you mean?

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/apps/ViewApps.java (hunk 1 above)
Comment (plata) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/apps/ViewApps.java]: How?

**Comment by plata** [comment]
Forget about it. I don't know regarding classpath.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/apps/ViewApps.java (hunk 1 above)
Comment (plata) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/apps/ViewApps.java]: Done.

**Pull request merged**